### ISSUE TITLE
Incorrect usage reported for the previous month.

### ISSUE THREAD

**Issue by piotrprzybylski** [cloudfoundry-incubator/cf-abacus#182]
The usage reported for the previous month at the month boundary does not seem to be correct. 
Setup: 

```
SLACK: 2D
resource: linux-container 
```

Three submissions as follows:

- usage 1 start/end at 1446249600000 (Sat Oct 31 2015 00:00:00 GMT+0000 (GMT)) 2GB - submitted at 10/31 02:00 GMT
- usage 2 start/end at 1446336000000 (Sun Nov 01 2015 00:00:00 GMT+0000 (GMT)) 2GB - submitted at 11/01 02:00 GMT [Artificial event inserted at the month boundary]
- usage 3 start/end at 1446415200000 (Sun Nov 01 2015 22:00:00 GMT+0000 (GMT)) 1GB - submitted at 11/01 23:00 GMT

```
  "usage": [
    {
      "start": 1446249600000,
      "end": 1446249600000,
      "region": "us-south",
      "organization_id": "us-south:b3d7fe4d-3cb1-4cc3-a831-ffe98e20cf27",
      "space_id": "baeae239-f3f8-483c-9dd0-de5d41c38b6a",
      "consumer_id": "cloud-foundry-application:baaaaaaa-f3f8-483c-9dd0-de5d41c38b6a",
      "resource_id": "linux-container",
      "plan_id": "basic",
      "resource_instance_id": "cloud-foundry-application:baaaaaaa-f3f8-483c-9dd0-de5d41c38b6a",
      "measured_usage": [
        {
          "measure": "instance_memory",
          "quantity": 1073741824
        },
        {
          "measure": "running_instances",
          "quantity": 2
        }
      ]
    }
  ]
}

```

```
{
  "usage": [
    {
      "start": 1446336000000,
      "end": 1446336000000,
      "region": "us-south",
      "organization_id": "us-south:b3d7fe4d-3cb1-4cc3-a831-ffe98e20cf27",
      "space_id": "baeae239-f3f8-483c-9dd0-de5d41c38b6a",
      "consumer_id": "cloud-foundry-application:baaaaaaa-f3f8-483c-9dd0-de5d41c38b6a",
      "resource_id": "linux-container",
      "plan_id": "basic",
      "resource_instance_id": "cloud-foundry-application:baaaaaaa-f3f8-483c-9dd0-de5d41c38b6a",
      "measured_usage": [
        {
          "measure": "instance_memory",
          "quantity": 1073741824
        },
        {
          "measure": "running_instances",
          "quantity": 2
        }
      ]
    }
  ]
}

```

```
  "usage": [
    {
      "start": 1446415200000,
      "end": 1446415200000,
      "region": "us-south",
      "organization_id": "us-south:b3d7fe4d-3cb1-4cc3-a831-ffe98e20cf27",
      "space_id": "baeae239-f3f8-483c-9dd0-de5d41c38b6a",
      "consumer_id": "cloud-foundry-application:baaaaaaa-f3f8-483c-9dd0-de5d41c38b6a",
      "resource_id": "linux-container",
      "plan_id": "basic",
      "resource_instance_id": "cloud-foundry-application:baaaaaaa-f3f8-483c-9dd0-de5d41c38b6a",
      "measured_usage": [
        {
          "measure": "instance_memory",
          "quantity": 1073741824
        },
        {
          "measure": "running_instances",
          "quantity": 1
        }
      ]
    }
  ]
}
```

Retrieve report at 2015-11-03T05:00:57Z which is after the SLACK period of 2 days 

```
curl -v http://localhost:9088/v1/metering/organizations/us-south%3Ab3d7fe4d-3cb1-4cc3-a831-ffe98e20cf27/aggregated/usage
```

Since there were two instances of the application with 1GB of memory running in October for 24 hours, 
the aggregated usage should be 24 hours * 2 instances - 48 hours or 172800000 milliseconds. 

But the report shows the following month usage

```
                [
                  {
                    "quantity": {
                      "consumed": 158400000,
                      "consuming": 1,
                      "since": 1446415200000
                    },
                    "cost": {
                      "burned": 22176,
                      "burning": 0.00014,
                      "since": 1446415200000
                    },
                    "summary": 75.0578675,
                    "charge": 0.01050810145
                  },
                  {
                    "quantity": {
                      "consumed": 0,
                      "consuming": 4,
                      "since": 1446249600000
                    },
                    "cost": {
                      "burned": 0,
                      "burning": 0.00056,
                      "since": 1446249600000
                    },
                    "summary": 308.23147,
                    "charge": 0.0431524058
                  }
                ]
```

The full report is attached

[scenario-1-linux.json.txt](https://github.com/cloudfoundry-incubator/cf-abacus/files/47939/scenario-1-linux.json.txt)

**Comment by BetaFood** [comment]
There's a couple of things to look at here.

1) Based upon the usage records you posted, there seems to be an issue inside of the aggregation. The previous month(October in this case) is somehow marking itself as consuming 4 despite no further records for October were submitted (it should be 2).

2) For consumed, it makes sense for it to be 0 in October since there was no additional usage to accumulate against in October. The summary in this case would play the role of providing the correct consumed value.
However, the summary in this example does not check to see if the current time falls outside of the current usage's window. For example, the summary function is always taking the difference of time instead of taking the difference for the end of October to calculate its value. This is true of the charge function as well.
I believe one way to go about this is to pass the current window's timescale(second, minute, hour, day, month) into the summary/charge functions to allow the resource config to properly calculate a charge/summary within a certain window.

3) Based upon the full report, it seems like the previous day slack isn't being populated on a month change despite the usage fulling within that slack window.

**Comment by BetaFood** [comment]
To clarify on what i said on point two in my last comment regarding summary/charge functions.

Given a window of(based on the example linux-container):
```javascript
[
  [
    {
      "quantity": 0
    }
  ],
  [
    {
      "quantity": 0
    }
  ],
  [
    {
      "quantity": 0
    }
  ],
  [
    {
      "quantity": {
        "consuming": 2,
        "consumed": 0,
        "since": 1448409600000
      }
    }
  ],
  [
    {
      "quantity": {
        "consuming": 2,
        "consumed": 0,
        "since": 1448409600000
      }
    }
  ]
]
```

Assuming no further usage submissions, if a report for this is called on 1449100800000, summary/charge will be called on every window.

If we used the day window in this case for calculating the summary, we would get:
(1449100800000 - 1448409600000) * 2 / 3600000 = 384.
Using the month window, the summary will result in the same value.

Ideally, the summary function should be able to tell when t has passed beyond its time threshold.
For instance, using the day window again.
1448409600000 is November 25th. Instead of using 1449100800000(December 3rd), it may rather calculate against 1448496000000(November 26th) to get (1448496000000 - 1448409600000) * 2 / 3600000 = 48 for the whole day of November 25th. This means that the summary for that day window would be for the usage of that day rather than the usage for that day up to December 3rd.
Likewise, if we use the month window.
Instead of using 1449100800000(December 3rd), it may want to use 1448928000000(December 1st) to get (1449100800000- 1448928000000) * 2 / 3600000 = 96 for the month of November.

In short, the summary/charge functions may want to be aware on which window(second, minute, hour, day, month) window they're working on such that a resource provider is able to correctly set boundaries for calculation instead of always using the query time, such that the summary/charge may bleed over into a different window.

**Comment by jsdelfino** [comment]
If I understand correctly, that means that the summarize and charge functions need to be passed two times 'from', and 'to' (the boundaries of the given time window) instead of just 't' (the report time).

Thoughts?

**Comment by BetaFood** [comment]
I think that's a good approach. I'll add from/to to the summarize and charge functions that will represent the start and end of the window for a particular slack and time scale.

For this specific case, it would just need to choose to calculate against to rather than the query time.

**Comment by piotrprzybylski** [comment]
I also agree, the calculations should be independent of slack setting.

**Comment by BetaFood** [comment]
These problems should now be fixed with b8802b054ec6f352387f730594e77c003865dab0 a269ff60ce41a085e53b52ea32eae58b07470cc6 500a0ae89999137c65ce95238a686297fa88c7db 148c90545ff012ffd13d7d7b19d089a0a61e7fa5
